core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala

```diff
@@ -231,16 +232,17 @@ private[delta] class ConflictChecker(
       // Fail if files have been deleted that the txn read.
       val readFilePaths = currentTransactionInfo.readFiles.map(
         f => f.path -> f.partitionValues).toMap
-      val deleteReadOverlap = winningCommitSummary.removedFiles
+      val deleteReadOverlap = winningCommitSummary.changedDataRemovedFiles
```

Comment:
Ignoring the dataChange=false files can cause serializability issues. A txn which is trying to commit relies on the compacted files and so it indirectly relies on new files created by OPTIMIZE.

Consider the following scenario:

    
    Suppose table has 2 files initially: f1, f2.
    t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
    t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
    t2 -------------   TXN-1 Commits:   	            Tries to commit [AddFile-F11, RemoveFile-F1] (This would have failed in master because of
                                                        ConcurrentDeleteReadException around file f1) (This commits successfully with new changes.)

At time t2 when TXN-1 tries to commit, it depends on file f1 and file f1 shouldn't change concurrently. In this case f1 was just rewritten to f3 by the concurrent OPTIMIZE and so this way we have a transitive dependency on f3 which we are missing here.

If we allow such TXN-1 to commit here, we could have issues like this:

    Suppose table has 2 files initially: f1, f2.
    t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
    t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
    t1.5 ------        TXN-2 starts and commits:        It deletes F3. i.e. it commits RemoveFile-F3
    t2 -------------   TXN-1 Commits:   	            Tries to commit F11 and it works as it doesn't conflict with OPTIMIZE (with new logic) and it doesn't conflict with TXN-2 (TXN-2 just deleted F3 which TXN-1 was not reading directly).

Here at time t2, we will allow TXN-1 to commit successfully (after proposed changes) although TXN-1 can't be serialized after TXN-2. So ideally TXN-1 should have failed.

Reply:
@prakharjain09 - this makes sense to me. thanks for reviewing this PR and lending a hand.

@sezruby do you understand + agree with this counter-example?

Reply:
Yes I understand the case. But OPTIMIZE-APPEND is common case for many users.
If there are frequent append jobs to a partition 24/7, we cannot run optimize for the partition.
Can we add additional check for the case? 

We have readPredicates used in `checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn`  and we can detect if there is a data change commit in each partition based on it. 
I think partition level detection could be feasible.  (I was thinking tracking optimized files using BF, but checking partition would be simple/enough)

What do you think about this approach? keep the current change + partition level dataChange='true' detection

Reply:
@scottsand-db @prakharjain09 I updated the PR with additional check & test for the case. Could you have a look at it? 
(I haven't addressed comment related change yet)

Reply:
@sezruby Could you give some example scenarios (like the one in this [comment](https://github.com/delta-io/delta/pull/1305#discussion_r986304404)) which shows the behavior before your changes and after your changes.

I think the main issue that you are trying to fix is: When Optimize commits, it affects a concurrent Append/Update/Delete/Merge query that is in progress and wants to commit due to checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn check? Is my understanding correct?

Reply:
I'd like to support OPTIMIZE - APPEND scenario only. 
Other operation should fail as we cannot track modified/removed rows from optimized files and it's acceptable.
Others with RemoveFile will conflict with optimize commit in checkForDeletedFilesAgainstCurrentTxnDeletedFiles and fail.

```
Suppose table has 2 files initially: f1, f2.
t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
t1.5 ------        TXN-2 starts and commits:        It deletes F3. i.e. it commits RemoveFile-F3
t2 -------------   TXN-1 Commits:   	            Tries to commit F11, but it fails in checkForDeletedFilesAgainstCurrentTxnReadFiles, as winningCommitSummary.changedDataRemovedFiles with readPredicate filter is not empty.
```

In current implementation, 
`checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn` checks newly added files using read predicates.
`checkForDeletedFilesAgainstCurrentTxnReadFiles` checks newly removed files only for readFiles of the current transaction.

I added read predicate check in `checkForDeletedFilesAgainstCurrentTxnReadFiles` so that we could detect removed files other than readFiles.

Reply:
I think this (APPEND_start / OPTIMIZE - DELETE / APPEND_end) is only case that we need to check "RemoveFiles" other than readFiles. Using data filter, there could be other files not in readFiles list (but in the same partition), so the additional check would make more unnecessary(?) conflicts. (e.g. fileA only contains id=1, fileB only contains id=2 only and with a data filter id=1, fileB is not in readFiles)

Do you have any other idea how we could support OPTIMIZE-APPEND?

Reply:
Hi @sezruby - things are getting a little complicated here :) 

Can we transition this to a more high-level discussion on the PR (not on this code) explaining the higher-level explanation/reasoning for this PR.

Inside of `checkForDeletedFilesAgainstCurrentTxnReadFiles` you seem to be checking the read predicates ... against a winningCommit's deleted files ....

@prakharjain09 is this valid?

Reply:
@sezruby can you please clarify what you mean by APPENDs?

`Reads F1 and Wants to Write F11`

Why are you reading first before these appends? Why not do BLIND appends?